src/Stratis.Bitcoin.Features.BlockStore/AddressIndexing/AddressIndexer.cs

```diff
@@ -510,75 +500,60 @@ public bool IsSynced()
         }
 
         /// <inheritdoc />
-        public Money GetAddressBalance(string address, int minConfirmations = 1)
+        public AddressBalancesResult GetAddressBalances(string[] addresses, int minConfirmations = 1)
         {
-            if (this.addressIndexRepository == null)
-            {
-                this.logger.LogTrace("(-)[NOT_INITIALIZED]");
-                throw new IndexerNotInitializedException();
-            }
+            var (isQuerable, reason) = this.IsQuerable();
+            if (!isQuerable)
+                return AddressBalancesResult.RequestFailed(reason);
 
-            if (!this.IsSynced())
-            {
-                this.logger.LogTrace("(-)[NOT_SYNCED]");
-                throw new OutOfSyncException();
-            }
+            var result = new AddressBalancesResult();
 
             lock (this.lockObject)
             {
-                AddressIndexerData indexData = this.addressIndexRepository.GetOrCreateAddress(address);
-                if (indexData == null)
+                foreach (var address in addresses)
                 {
-                    this.logger.LogTrace("(-)[NOT_FOUND]");
-                    return null;
-                }
+                    AddressIndexerData indexData = this.addressIndexRepository.GetOrCreateAddress(address);
+                    if (indexData == null)
+                    {
+                        this.logger.LogDebug("{0} was not found in the indexer.", address);
+                        result.Balances.Add(new AddressBalanceResult(address, new Money(0)));
+                        continue;
+                    }
 
-                long balance = 0;
+                    long balance = 0;
 
-                int maxAllowedHeight = this.consensusManager.Tip.Height - minConfirmations + 1;
+                    int maxAllowedHeight = this.consensusManager.Tip.Height - minConfirmations + 1;
```

Comment:
The intended design for the indexer is that any balance changes older than maxReorg get compacted into a single entry for storage efficiency. Therefore, we should probably check if the `minConfirmations` required is greater than maxReorg and return an error in that case, as I don't think we can return a meaningful result. What do you think?

Reply:
Good point but perhaps we should do this another PR?